src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -278,31 +281,99 @@ def process_input(conf):
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
 
-        # cfn points to the assembly file outside of the container
-        # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
-        # workdir is shared between the container and the external file system
-        # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+        if retrieve_sample:
+            _download_sample_data(conf, upa)
 
-        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
-        # upa file is downloaded to cfn
-        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
 
-        # each upa in output_dir as a seperate directory
-        dstd = os.path.join(conf.pth, upa)
-        os.makedirs(dstd, exist_ok=True)
+def _download_sample_data(conf, upa):
+    # retrieve sample data from sample service and save to file
 
-        dst = os.path.join(dstd, f"{upa}.fa")
-        # Hard link .fa file from job_dir to output_dir in WS
-        os.link(cfn, dst)
+    dstd = os.path.join(conf.pth, upa)
+    os.makedirs(dstd, exist_ok=True)
+    sample_file = os.path.join(dstd, f"{upa}.sample")
 
-        metafile = os.path.join(dstd, f"{upa}.meta")
-        # save meta file with relevant object_info
-        with open(metafile, "w", encoding="utf8") as json_file:
-            json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+    if os.path.isfile(sample_file):
+        print(f"Skip downloading sample for {upa} as it already exists")
+        return
 
-        print("Completed %s" % (upa))
+    # retrieve data links associated with upa
+    links_ret = _post_sample_service(conf.token,
+                                     conf.sample_url,
+                                     "get_data_links_from_data",
+                                     {"upa": upa.replace("_", "/")})
+    data_links = links_ret['links']
+    if not data_links:
+        print(f"No sample data links found for {upa}")
+        return
+
+    # there should only be one data link for each upa
+    if len(data_links) != 1:
+        raise ValueError(f"Expected 1 data link for {upa}, got {len(data_links)}")
+
+    # retrieve sample data and save to file
+    sample_id = data_links[0]['id']
+    sample_ret = _post_sample_service(conf.token,
+                                      conf.sample_url,
+                                      "get_sample",
+                                      {"id": sample_id})
+
+    with open(sample_file, "w", encoding="utf8") as json_file:
+        json.dump(sample_ret, json_file, indent=2)
```

Comment:
Before writing the sample needs to be flattened into a single level key-value map. We should error out if there's a key other than "value" in the sample field map (except for latitude and longitude, which also have "units") and put the latitude and longitude keys into the common names file for use in indexing and aggregations in the service.

We may need to switch lat/long to a single field like `"location": [<lat>, <long>]` but I'm not sure about that yet.

What I don't want to have happen is that samples for other collections have different fields and we silently ignore them, hence the errors

Reply:
Do it here? I was thinking about do it in the parser.

Reply:
To my mind, the downloaders should do any special data munging that depends on the source, and put it into a common format for the parser to handle. Keep the parser as simple as possible to my mind

Reply:
Also means that we fail early if there's an unexpected sample format

Reply:
okay. that's fine. Do you know if we already have this logic implemented somewhere?

Reply:
I also have no idea how samples data will be used. Just parse something like below from `meta_controlled` and `meta_user` and save the dict to the sample file? If we change parsing logic down the road, we will just redownload samples from sample service and re-parse? 
`
{
'depth_bgs': 299,
'country': 'USA',
'sesar:size': 1
}
`

Reply:
yeah, that's what I was thinking. The sample data will be used in the exact same way the genome attribs data is used, except a different table, at least for the first cut.

Reply:
Let's not bother with `meta_user` though

Reply:
> If we change parsing logic down the road, we will need to redownload all samples from sample service and re-parse?

Good point - let's save the original and the mutated version. They're tiny so NDB

Reply:
👍 added logic to retrieve node data.

Reply:
:+1: